### DISCUSSION TITLE
Tree and Items Virtualization Questions

### DISCUSSION THREAD

**Discussion by fitdev** [AvaloniaUI/Avalonia#12436] [Q&A]
The documentation seems to be a bit sketchy at the moment, so decided to ask to clarify a few things here:

1. As per [docs](https://docs.avaloniaui.net/docs/next/stay-up-to-date/upgrade-from-0.10#itemscontrol) virtualization is enabled by setting `ItemsControl.ItemsPanel` to `VirtualizingStackPanel`. It does not seem to work for `TreeView` though as it only sets the panel for the first level in the hierarchy, subsequent levels continue to use regular `StackPanel`:

```cs
public partial class HierarchicalTest1 : Window {
  public HierarchicalTest1() {
    InitializeComponent();
    MyTreeView.ItemsPanel = new VirtualTreePanelTemplate();
  }
}
public class VirtualTreePanelTemplate : ITemplate<Panel?> {
  public Panel? Build() => new VirtualizingStackPanel() { Orientation = Avalonia.Layout.Orientation.Vertical };
  object? ITemplate.Build() => Build();
}
```

![Screenshot_20230804_110659](https://github.com/AvaloniaUI/Avalonia/assets/8166847/b2fbc5ab-6bca-4470-8eab-fed1ae461ef6)

Is this a bug? Or did I do something wrong?

2. At least for flat list controls (non-tree), is virtualization as simple as just replacing `StackPanel` with `VirtualizingStackPanel`, or is there something more to it? How do I then control the scrollbars, i.e. give hints to how many more items exist above and below the current view and what their sizes should be?

**Comment by timunie** [comment]
TreeView is not able to virtualized. Try Avalonia.TreeDataGrid.

**Reply by fitdev** [reply]
I see. As a side question, if I want flat list-based control (flat) with a _tile-based layout_ that _supports virtualization_ then this is currently not possible, since there's just `VirtualizingStackPanel` but no `VirtualizingWrapPanel` which I presume is what would be needed to enable tiled mode?

**Reply by fitdev** [reply]
I also noticed that if I use `VirtualizingStackPanel` even with regular `ListBox`, when I hide _not yet "loaded"/displayed_ items by setting `ListBoxItem.IsVisible` to `false` (via binding in a style), then when I scroll, those items remain visible, despite their visibility being set to false. This does not happen with regular panel. Is this a bug?

**Reply by fitdev** [reply]
It seems there is some kind of rush condition then between when `ListBoxItem`s are being actualized by the `VirtualizingStackPanel` and when a style is being applied to those items, setting their `IsVisible` to `false` - for some items it works for some it does not.
